### DISCUSSION TITLE
endpoints

### DISCUSSION THREAD

**Discussion by vishalkadam18** [apache/skywalking#11791] [Q&A] [answered]
Is this apm tool use for more than 200+ endpoints

**Comment by wu-sheng** [comment]
Yes, it is a common case for thousands of endpoints. We don't hold any of them in your application memory, all data are collected and reported based on your traffic. The memory will be released after the report. So, your application is safe.

**Comment by vishalkadam18** [comment]
1. how can i increase my endpoints limits i have increase from frontend but that was not working when we across limits then that not showing any endpoints??
2. we have to increase limits from backend also??

**Reply by wu-sheng** [reply]
I don't think we open that option, although the query limit exists from the protocol level. 
<1> What do you mean `not working`?
<2> Duplicated with https://github.com/apache/skywalking/discussions/11780

**Reply by vishalkadam18** [reply]
<1> it means that when we cross limit of endpoints then this type blank is coming
![Screenshot from 2024-01-18 12-24-10](https://github.com/apache/skywalking/assets/115125925/b21c4cc8-40aa-45ff-a416-3cab201ea50e)

**Reply by wu-sheng** [reply]
No, that is not expected I think. The query of this page is applied for a random 20 only, depending on the storage

**Comment by vishalkadam18** [comment]
can we increase endpoints limit???

**Reply by wu-sheng** [reply]
From the API level, yes, but not available on UI configuration.
So you need to change codes to make this happens.

**Comment by vishalkadam18** [comment]
it means we have to change all code or make some changes on UI level

**Reply by wu-sheng** [reply]
Yes. You could send a pull request to make that changed on upstream, and benefit for everyone.

**Comment by vishalkadam18** [comment]
I have done all things to increase endpoints limits from UI level but still is not working when we cross 52+ endpoints then this type of error occurs: 
Error: More than 15,000 'grammar' tokens have been presented. To prevent Denial Of Service attacks, parsing has been cancelled.

**Reply by wu-sheng** [reply]
AFAIK, this is controlled by GraphQL Parsing.properties. We don't have set them, so, I assume they are default values.

```properties
ParseCancelled.full=More than {0} ''{1}'' tokens have been presented. To prevent Denial Of Service attacks, parsing has been cancelled.
ParseCancelled.tooDeep=More than {0} deep ''{1}'' rules have been entered. To prevent Denial Of Service attacks, parsing has been cancelled.
ParseCancelled.tooManyChars=More than {0} characters have been presented. To prevent Denial Of Service attacks, parsing has been cancelled.
```

**Comment by vishalkadam18** [comment]
when i increase endpoints limits 20+ on UI level then this type of error occurs
Error: maximum query complexity exceeded 1080 > 1000

**Reply by wu-sheng** [reply]
This is limited by OAP applciation.yml

```
    # Maximum complexity allowed for the GraphQL query that can be used to
    # abort a query if the total number of data fields queried exceeds the defined threshold.
    maxQueryComplexity: ${SW_QUERY_MAX_QUERY_COMPLEXITY:3000}
```

**Reply by wu-sheng** [reply]
It was 1000, and increased to 3000 already. You may not be using the latest.

**Reply by vishalkadam18** [reply]
after this changes then also giving same issue but i am unable to run new latest source code is there any way to run latest source code

**Reply by wu-sheng** [reply]
We are running from source codes every day. I am not sure what do you mean can't.

**Reply by vishalkadam18** [reply]
i mean that i am unable to run latest source code in my local machine when we clone from git then there are some command for run the project that is not working is there any solution or please provide me link of latest git source code where i can use this features...

**Reply by wu-sheng** [reply, marked as answer]
You should read docs, dev stage is more relying on your own.

https://skywalking.apache.org/docs/main/next/en/guides/how-to-build/

**Reply by vishalkadam18** [reply]
when i have increase limit more that 20+ of endpoints from UI level then all endpoints is not visible and giving this type of error
Error: maximum query complexity exceeded 1440 > 1000
![Screenshot from 2024-01-18 12-24-10](https://github.com/apache/skywalking/assets/115125925/b21c4cc8-40aa-45ff-a416-3cab201ea50e)

after doing this changes also

This is limited by OAP applciation.yml

    # Maximum complexity allowed for the GraphQL query that can be used to
    # abort a query if the total number of data fields queried exceeds the defined threshold.
    maxQueryComplexity: ${SW_QUERY_MAX_QUERY_COMPLEXITY:3000}

**Reply by vishalkadam18** [reply]
In latest source code can we get more endpoints more then 20+ or not????

**Reply by vishalkadam18** [reply]
i have taken latest code but then also when we increase this then where should we make much changes to take much more endpoints
eg. metadataQueryMaxSize: ${SW_STORAGE_ES_QUERY_MAX_SIZE:3000} this to this 
metadataQueryMaxSize: ${SW_STORAGE_ES_QUERY_MAX_SIZE:5000}

after this changes then also this error occurs
error:maximum query complexity exceeded 3744 > 3000

**Comment by vishalkadam18** [comment]
can we replace old version UI on new version it is possible?????
